app/src/main/java/com/github/se/project/viewmodels/AuthenticationViewModel.kt

```diff
@@ -0,0 +1,140 @@
+        val bytes: ByteArray = ranNonce.toByteArray()
+        val md: MessageDigest = MessageDigest.getInstance("SHA-256")
+        val digest: ByteArray = md.digest(bytes)
+        val hashedNonce: String = digest.fold("") { str, it -> str + "%02x".format(it) }
+
+        // Set up Google ID option with necessary parameters
+        val googleIdOption: GetGoogleIdOption =
+            GetGoogleIdOption.Builder()
+                .setFilterByAuthorizedAccounts(
+                    false) // To give the user the option to choose from any Google account on their
+                // device, not just the ones they've used with your app before.
+                .setServerClientId(
+                    context.getString(
+                        R.string
+                            .web_client_id)) // This is required to identify the app on the backend
+                // server.
+                .setNonce(
+                    hashedNonce) // A nonce is a unique, random string used to ensure that the ID
+                // token received is fresh and to prevent replay attacks.
+                .setAutoSelectEnabled(
+                    true) // Which allows the user to be automatically signed in without additional
+                // user interaction if there is a single eligible account.
+                .build()
+
+        // Create a credential request with the Google ID option
+        val request: GetCredentialRequest =
+            GetCredentialRequest.Builder().addCredentialOption(googleIdOption).build()
+
+        // Get the credential result from the Credential Manager
+        val result = credentialManager.getCredential(context, request)
+        val credential = result.credential
+
+        // Check if the received credential is a valid Google ID Token
+        if (credential is CustomCredential &&
+            credential.type == GoogleIdTokenCredential.TYPE_GOOGLE_ID_TOKEN_CREDENTIAL) {
+          // Extract the Google ID Token credential
+          val googleIdTokenCredential = GoogleIdTokenCredential.createFrom(credential.data)
+          // Create an auth credential using the Google ID Token
+          val authCredential =
+              GoogleAuthProvider.getCredential(googleIdTokenCredential.idToken, null)
+          // Sign in with Firebase using the auth credential
+          val authResult =
+              firebaseAuth
+                  .signInWithCredential(authCredential)
+                  .await() // .await() -> allows the coroutine to wait for the result of the
+          // authentication operation before proceeding.
+          // Send the successful result
+          trySend(
+              Result.success(
+                  authResult)) // Is used to send the result of the Firebase sign-in operation to
+          // the Flow's collectors.
+        } else {
+          // Throw an exception if the credential type is invalid
+          throw RuntimeException("Received an invalid credential type.")
```

Comment:
It might be helpful to log detailed error messages into the logcat as follow: catch (e: Exception) {
    Log.e("AuthenticationViewModel", "Error during Google Sign-In", e)
}
same for the catch here under